components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/serialize/BrowserStateReader.kt

```diff
@@ -60,76 +57,62 @@ class BrowserStateSerializer {
     }
 }
 
-/**
- * Writes [BrowserState] to [JsonWriter].
- */
-private fun JsonWriter.state(
-    state: BrowserState
-) {
+@Suppress("ComplexMethod")
+private fun JsonReader.browsingSession(
+    engine: Engine,
+    restoreSessionId: Boolean = true,
+    restoreParentId: Boolean = true
+): RecoverableBrowserState? {
     beginObject()
 
-    name(Keys.VERSION_KEY)
-    value(VERSION)
-
-    name(Keys.SELECTED_TAB_ID_KEY)
-    value(state.selectedTabId)
-
-    name(Keys.SESSION_STATE_TUPLES_KEY)
+    var version = 1 // Initially we didn't save a version. If there's none then we assume it is version 1.
+    var tabs: List<RecoverableTab>? = null
+    var selectedIndex: Int? = null
+    var selectedTabId: String? = null
 
-    beginArray()
-
-    state.tabs.filter { !it.content.private }.forEachIndexed { _, tab ->
-        tab(tab)
+    while (hasNext()) {
+        when (nextName()) {
```

Comment:
Hm we're having a strange crash caused by `nextName`: https://sentry.prod.mozaws.net/operations/firefox-nightly/issues/10295333/

This really shouldn't happen according to docs and after checking `hasNext` 🤔

Reply:
I can't currently open that. Can you post part of the stack? In what code is it happening? (Since this here hasn't landed yet)

Reply:
In our `BrowserStateSerializer`:

```
java.lang.IllegalStateException: Expected a name but was END_OBJECT
    at android.util.JsonReader.nextName(JsonReader.java:390)
    at mozilla.components.browser.session.storage.BrowserStateSerializerKt.access$tab(BrowserStateSerializer.kt:36)
    at mozilla.components.feature.tab.collections.adapter.TabAdapter.restore(TabAdapter.kt:7)
    at org.mozilla.fenix.home.sessioncontrol.DefaultSessionControlController.handleCollectionOpenTabClicked(SessionControlController.kt:8)
    at org.mozilla.fenix.home.sessioncontrol.SessionControlInteractor.onCollectionOpenTabClicked(SessionControlInteractor.kt:1)
    at org.mozilla.fenix.home.sessioncontrol.viewholders.-$$LambdaGroup$js$CmLMMZGdtRGl-qY2EQsQLG5w5hQ.onClick(java-style lambda group:1)
    at android.view.View.performClick(View.java:5646)
    at android.view.View$PerformClick.run(View.java:22473)
    at android.os.Handler.handleCallback(Handler.java:761)
    at android.os.Handler.dispatchMessage(Handler.java:98)
    at android.os.Looper.loop(Looper.java:156)
    at android.app.ActivityThread.main(ActivityThread.java:6517)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
```

Reply:
This happens in latest Nightly too and I can't find any usage of `nextName` that doesn't check for `hasNext`...so it's  weird.

Reply:
Does it happen often? Would we want to block this on it?

Reply:
Not often 85 events / 28 users. Nothing obvious in the metrics. All kinds of different Android versions and devices.

Reply:
One thing I noticed that could cause this is that we don't guard against concurrent access in tab-collections. In `SessionStorage` and everywhere else we use `AtomicFile`s we guard using a lock, but we don't do that for collections? This seems like a bug? We can fix / investigate separately.

Reply:
Yeah, sounds like something we could investigate. Not sure how that would actually happen when using the app, but since it is low in volume maybe some users were able to pull it off.

Reply:
@csadilek @pocmo Please see https://github.com/mozilla-mobile/fenix/issues/17219 - except of the "nextName" line the stack trace looks very similar. And in the Firefox support we also got a similar report from another user. In both cases the crash only happens if you try to access a collection that was created before a certain build with a newer build and it's not reproducible with newly created collections.

Reply:
@cadeyrn I think what users have been experiencing here most likely is https://github.com/mozilla-mobile/android-components/issues/9292 which also lines up with the stacktrace in the ticket you linked. This should already be fixed in latest Nightly.